cohttp-curl-lwt/src/cohttp_curl_lwt.ml

```diff
@@ -0,0 +1,247 @@
+module Sink = struct
+  type _ t = String : string t | Discard : unit t
+
+  let string = String
+  let discard = Discard
+end
+
+module Source = struct
+  type t = Empty | String of string
+
+  let empty = Empty
+  let string s = String s
+
+  let to_curl_callback t =
+    match t with
+    | Empty -> fun _ -> ""
+    | String s ->
+        let len = String.length s in
+        let pos = ref 0 in
+        fun max_asked ->
+          if !pos >= len then ""
+          else
+            let chunk_len = min (len - !pos) max_asked in
+            let res = String.sub s !pos chunk_len in
+            pos := !pos + chunk_len;
+            res
+end
+
+module Context = struct
+  type t = {
+    mt : Curl.Multi.mt;
+    wakeners : (Curl.t, Curl.curlCode Lwt.u) Hashtbl.t;
+    all_events : (Unix.file_descr, Lwt_engine.event list) Hashtbl.t;
+    mutable timer_event : Lwt_engine.event;
+  }
+
+  let create () =
+    (* Most of this is taken from https://github.com/ygrek/ocurl/blob/master/curl_lwt.ml *)
```

Comment:
this should preserve copyright from https://github.com/ygrek/ocurl/blob/master/COPYING